common/emplace_result.h

```diff
@@ -0,0 +1,64 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_COMMON_EMPLACE_RESULT_H_
+#define CARBON_COMMON_EMPLACE_RESULT_H_
+
+#include <type_traits>
+#include <utility>
+
+namespace Carbon {
+
+// A utility to use when calling an `emplace` function to emplace the result of
+// a function call. Expected usage is:
+//
+//   my_widget_vec.emplace_back(EmplaceResult([&] {
+//     return ConstructAWidget(...);
+//   }));
+//
+// In this example, the result of `ConstructAWidget` will be constructed
+// directly into the new element of `my_widget_vec`, without performing a copy
+// or move.
+//
+// This works by providing a conversion function on `EmplaceResult` that
```

Comment:
The part of this explanation that was missing for me was:
```cc
// This works by making the emplace parameter type be an `EmplaceResult`, instead of
// being the type `DestT` being stored in the container. The `EmplaceResult` is moved
// through to the construction site, avoiding any move of a `DestT`, and then constructs
// `DestT` for the first time inside the call to construct the `DestT` member in the
// container's storage. 
```
Maybe this can be fit into here somewhere?

Reply:
Rephrased and explicitly pointed out that the argument to the emplace function is an `EmplaceResult` instance. Does this work better for you?